src/main/java/com/iota/iri/service/spentaddresses/impl/SpentAddressesServiceImpl.java

```diff
@@ -140,4 +124,113 @@ private boolean isBundleValid(Hash tailHash) throws Exception {
                 BundleValidator.validate(tangle, snapshotProvider.getInitialSnapshot(), tailHash);
         return (CollectionUtils.isNotEmpty(validation) && validation.get(0).get(0).getValidity() == 1);
     }
+
+    /**
+     *
+     * @param addressHash the address in question
+     * @param checkedAddresses a list of known unspent addresses, used to skip calculations
+     * @return {@code true} if address was spent from, else {@code false}
+     * @throws SpentAddressesException
+     * @see #wasAddressSpentFrom(Hash)
+     */
+    private boolean wasAddressSpentFrom(Hash addressHash, Set<Hash> checkedAddresses) throws SpentAddressesException {
+        if (addressHash == null) {
+            return false;
+        }
+
+        if (spentAddressesProvider.containsAddress(addressHash)) {
+            return true;
+        }
+
+        //If address has already been checked this session, return false
+        if (SafeUtils.isContaining(checkedAddresses, addressHash)){
+            return false;
+        }
+
+        //If address is either null or equal to the coordinator address
+        if (addressHash.equals(Hash.NULL_HASH) ||
+                addressHash.toString().equals(config.getCoordinator())) {
+            return false;
+        }
+
+
+        try {
+            Set<Hash> hashes = AddressViewModel.load(tangle, addressHash).getHashes();
+            int setSizeLimit = 100000;
+
+            //If the hash set returned contains more than 100 000 entries, it likely will not be a spent address.
+            //To avoid unnecessary overhead while processing, the loop will return false
+            if (hashes.size() > setSizeLimit){
+                checkedAddresses.add(addressHash);
+                return false;
+            }
+
+            for (Hash hash: hashes) {
+                TransactionViewModel tx = TransactionViewModel.fromHash(tangle, hash);
+                // Check for spending transactions
+                if (wasTransactionSpentFrom(tx)) {
+                    return true;
+                }
+            }
+
+        } catch (Exception e) {
+            throw new SpentAddressesException(e);
+        }
+
+        if (checkedAddresses != null) {
+            checkedAddresses.add(addressHash);
```

Comment:
you do `checkedAddresses.add()` before in an unsafe way  (L164, even though it's in a try), is this check actually necessary?

Reply:
Missed it in the review.
It is necessary.

However, if I implement your above comment than we can be sure that `checkedAddress` won't be null and thus no null check will be neccessary